### ISSUE TITLE
Invalid use of http error codes for exception messaging/handling

### ISSUE THREAD

**Issue by TuxCoder** [austrianredcross/stopp-corona-android#26]
You should not use 500 as a way to say the client has a invalid value or a backend service is not working.

`app/src/main/java/at/roteskreuz/stopcorona/model/api/ApiInteractor.kt`
```java
...
    private val generalExceptionMapper: (HttpException) -> Exception? = {
        when (it.code()) {
            HTTP_FORBIDDEN -> ApiError.Critical.AuthorizationError
            HTTP_GONE -> ApiError.Critical.ForceUpdate
            else -> null
        }
    }
...
    override suspend fun setInfectionInfo(infectionInfoRequest: ApiInfectionInfoRequest) {
        withContext(appDispatchers.IO) {
            dataPrivacyRepository.assertDataPrivacyAccepted()
            checkGeneralErrors(
                { httpException ->
                    when (httpException.code()) {
                        HTTP_FORBIDDEN -> SicknessCertificateUploadException.TanInvalidException
                        HTTP_INTERNAL_ERROR -> SicknessCertificateUploadException.BirthdayInvalidException
                        else -> generalExceptionMapper(httpException)
                    }
                },
                {
                    apiDescription.infectionInfo(infectionInfoRequest)
                }
            )
        }
    }
...    
    override suspend fun requestTan(mobileNumber: String): ApiRequestTan {
        return withContext(appDispatchers.IO) {
            dataPrivacyRepository.assertDataPrivacyAccepted()
            checkGeneralErrors({ httpException ->
                when (httpException.code()) {
                    HTTP_UNAUTHORIZED -> SicknessCertificateUploadException.PhoneNumberInvalidException
                    HTTP_INTERNAL_ERROR -> SicknessCertificateUploadException.SMSGatewayException
                    else -> generalExceptionMapper(httpException)
                }
            },
                {
                    apiDescription.requestTan(mobileNumber)
                })
        }
    }
...
```

Proposed solution:
Use HTTP400 `Bad request` for rest application form errors and the reason should be a field in a response json

When you have a Gateway error please use the 502 for "Bad Gateway".
So 500 is reserved for a internal error.

**Comment by mitsuhiko** [comment]
You are correct that it would be reasonable to change this, however this would require a backend service change first and in light of all the other more important tasks I believe this is unlikely to be changed for now.

**Comment by DJCrashdummy** [comment]
just my 2 cents: i would rather consider this a `bug` than an `enhancement`...

**Comment by MovGP0** [comment]
I agree. this is a bug in the backend.

**Comment by akuckartz** [comment]
https://tools.ietf.org/html/rfc7231#section-6.6

502 is more specific than 500. If 502 is correct then 500 would be correct as well.

**Comment by TuxCoder** [comment]
@akuckartz it true, by the standard you can use both.

But when you want to different between an "internal server error" (what is normaly a exception in the server side) and and not working gateway (sms token/ crypto token verifyer/...) and show a proper error message to the user you need two different states.

This can be ether 500/502 as error code or a proper response like
`500` - `{'errors':{'internal': 'GATEWAY_ERROR'}}`
As we provide more information as only the error code we can different between different types of errors.
Warning: need to be translated locally, can also done server side by a Language header send with the request)

How is the proper way depend on many things, but using a 500 as invalid gateway is mixing for sure two different error messages.


Probably it makes sense to define the api in a abstract language like `swagger.io` or similar for easier understanding and have a more defined way, what kind of error happen when.

**Comment by TuxCoder** [comment]
The new API looks better, like proper REST API should look like, I think its not finished yet, but for everyone who is interessted the openapi json from the current public api from the RCA Backend:

https://gist.github.com/TuxCoder/1fa962b97357beb3b1918172cf045a65

you can look easy to it with any openapi 2.0 viewer like https://editor.swagger.io/ .
